keps/sig-api-machinery/4568-resilient-watchcache-initialization/README.md

```diff
@@ -0,0 +1,464 @@
+to retrieve the object from cache, in which case we simply avoid passing the
+current object down and it's then retrieved from etcd. We didn't observe nor
+heard about this causing real issues and we're not going to change this behavior
+as part of this KEP.
+
+For watch requests, unless it's a consistent watch request (setting RV="") - which
+isn't really used by none of our core components, it is always handled by the
+watchcache. For all such requests, watch is hanging until watchcache finishes
+initialization. The problem here is that for clusters with a large amount of data
+of a given type, initialization (being effectively a list from etcd) can take
+tens of seconds or even above a minute and in the meantime this watch is consuming
+seats of API Priority&Fairness. In many cases at most tens of such watches (which
+is a relatively small number in clusters with thousands of nodes) can effectively
+completely starve a given PriorityLevel and block any load on it (both write and
+read requests).
+
+For get/list requests that explicitly opt-in for being served from watchcache
+(namely those marked as `NotOlderThan` or `Any` in [the documentation]) - which
+amongst others is what Reflector/Informers framework is using in most of cases,
+if watchcache is not initialized there are two cases:
+- if ResourceVersion=0, these are blindly delegated to etcd
+- if ResourceVersion>0, they hang waiting for initialization similarly as watch
+requests
+
+The first case is particularly problematic for lists that hugely benefit from
+watchcache indexing (e.g. listing pods assigned to a given node). When served
+from watchcache, these requests are really efficient, but serving them from etcd
+requires reading a large amount of data, deserializing it, filtering out objects
+not matching selector and sending only those to the client. Passing too many of
+such requests down to etcd, can easily bring down the whole control plane.
+Additionally, in this case, when served from watchcache a LIMIT parameter is
+ignored and the whole result is returned in one go. If served from etcd, the
+LIMIT parameter is honored, so for large collections we end-up paging the
+response (which for large collections and limit=500 of reflector/informer
+frameworks can result in hundreds of consecutive requests for next pages).
+
+The second case is suffering from the same problem as watches, as those requests
+may effectively starve the whole PriorityLevel.
+
+In order to mitigate the above problem, we propose a couple different changes.
+
+[the documentation]: https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list
+
+#### Reduce the number of requests during initialization
+
+As a first step, we will try to reduce the number of requests that are even reaching
+kube-apiserver with unitialized watchcache. To achieve it, we will introduce a new
+PostStartHook that will wait for watchcache of all *builtin* resources to be initialized.
```

Comment:
Haven't finished yet, but this will also include KMS handling on the path for Secrets for instance.  While I'd like to be more efficient, I think we should  have a max-time that we're willing to wait.